File: src/main/java/fr/openent/viescolaire/service/TrombinoscopeService.java
package fr.openent.viescolaire.service;

import fr.openent.viescolaire.model.Trombinoscope.TrombinoscopeReport;
import io.vertx.core.AsyncResult;
import io.vertx.core.Future;
import io.vertx.core.Handler;
import io.vertx.core.http.HttpServerRequest;
import io.vertx.core.json.JsonObject;


public interface TrombinoscopeService {

    /**
     * get trombinoscope setting feature
     *
     * @param structureId   Structure Identifier
     * @param handler       Function handler returning data
     */
    void getSetting(String structureId, Handler<AsyncResult<Boolean>> handler);

    /**
     * set trombinoscope setting to toggle enable/disable (create if no exist)
     *
     * @param structureId   Structure Identifier
     * @param active        student Identifier
     * @param handler       Function handler returning data
     */
    void setSetting(String structureId, Boolean active, Handler<AsyncResult<JsonObject>> handler);

    /**
     * get trombinoscope info from a student
     *
     * @param structureId   Structure Identifier
     * @param studentId     student Identifier
     * @param handler       Function handler returning data
     */
    void get(String structureId, String studentId, final Handler<AsyncResult<JsonObject>> handler);

    /**
     * import trombinoscope
     *
     * @param structureId   Structure Identifier
     * @param path          Uploaded from file unzipped
     * @param request       HttpServerRequest
     * @param handler       Function handler returning data
     */
    void process(String structureId, String path, TrombinoscopeReport trombinoscopeReport,
                 final HttpServerRequest request, final Handler<AsyncResult<Void>> handler);

    /**
     * create trombinoscope
     *
     * @param structureId   Structure Identifier
     * @param studentId     Student identifier
     * @param pictureId     picture identifier
     * @param handler       Function handler returning data
     */
    void create(String structureId, String studentId, String pictureId, Handler<AsyncResult<JsonObject>> handler);

    /**
     * delete picture from a studentId
     *
     * @param structureId   Structure Identifier {@link String}
     * @param studentId     Student identifier {@link String}
     * @return Future       {@link Future} of {@link JsonObject} completed or failure
     */
    Future<JsonObject> deletePicture(String structureId, String studentId);
}

File: src/main/resources/public/modules/viescolaire/services/TrombinoscopeService.ts
import {ng} from 'entcore';
import http, {AxiosResponse} from 'axios';
import {IFailure, IReport} from '../models/trombinoscope';

export interface ITrombinoscopeService {
    importTrombinoscope(structureId: string, file: File): Promise<AxiosResponse>;

    updateTrombinoscope(structureId: string, studentId: string, file: File): Promise<AxiosResponse>;

    deleteTrombinoscope(structureId: string, studentId: string): Promise<AxiosResponse>;

    getFailures(structureId: string): Promise<Array<IFailure>>;

    linkTrombinoscope(structureId: string, studentId: string, pictureId: string): Promise<AxiosResponse>;

    setStructureSettings(structureId: string, useAvatar: boolean): Promise<AxiosResponse>;

    getStructureSettings(structureId: string): Promise<boolean>;

    getReports(structureId: string, limit: number, offset: number): Promise<Array<IReport>>;

    deleteFailuresHistory(structureId: string): Promise<AxiosResponse>;
}

export const trombinoscopeService: ITrombinoscopeService = {

    /**
     * Import trombinoscope zip file
     *
     * @param structureId   structure identifier
     * @param file          trombinoscope file
     */
    importTrombinoscope: async (structureId: string, file: File): Promise<AxiosResponse> => {
        const formData = new FormData();
        formData.append('file', file);
        return http.post(`/viescolaire/structures/${structureId}/trombinoscope`, formData);
    },

    /**
     * Update a student picture.
     *
     * @param structureId   structure identifier
     * @param studentId     student identifier
     * @param file          picture file
     */
    updateTrombinoscope: async (structureId: string, studentId: string, file: File): Promise<AxiosResponse> => {
        const formData: FormData = new FormData();
        const headers = {'headers': {'Content-type': 'multipart/form-data'}};

        formData.append('file', file);

        return http.put(`/viescolaire/structures/${structureId}/students/${studentId}/trombinoscope`, formData, headers);
    },

    /**
     * delete student picture (its trombinoscope).
     *
     * @param structureId   structure identifier
     * @param studentId     student identifier
     */
    deleteTrombinoscope: async (structureId: string, studentId: string) => {
        return http.delete(`/viescolaire/structures/${structureId}/students/${studentId}/trombinoscope`);
    },

    /**
     * Get trombinoscope failures.
     *
     * @param structureId   structure identifier
     */
    getFailures(structureId: string): Promise<Array<IFailure>> {
        return http.get(`/viescolaire/structures/${structureId}/trombinoscope/failures`)
            .then((res: AxiosResponse) => {
            return res.data.all;
        });
    },

    /**
     * Link failure picture to a student
     *
     * @param structureId   structure identifier
     * @param studentId     student identifier
     * @param pictureId     picture identifier
     */
    linkTrombinoscope: async (structureId: string, studentId: string, pictureId: string): Promise<AxiosResponse> => {
        return http.post(`/viescolaire/structures/${structureId}/students/${studentId}/trombinoscope`, { pictureId: pictureId});
    },

    /**
     * Set structure trombinoscope setting.
     * @param structureId   structure identifier
     * @param useAvatar     enable/diable use of trombinoscope images
     */
    setStructureSettings: async (structureId: string, useAvatar: boolean): Promise<AxiosResponse> => {
        return http.post(`/viescolaire/structures/${structureId}/trombinoscope/setting`, {active: useAvatar});
    },

    /**
     * Get structure trombinoscope settings.
     * @param structureId   structure identifier
     */
    getStructureSettings: (structureId: string): Promise<boolean> => {
        return http.get(`/viescolaire/structures/${structureId}/trombinoscope/setting`)
            .then((res: AxiosResponse) => {
                return res.data.active;
            });
    },

    /**
     * Get reports from structure.
     * @param structureId   structure identifier
     * @param limit
     * @param offset
     */
    getReports: (structureId: string, limit: number, offset: number): Promise<Array<IReport>> => {
        return http.get(`/viescolaire/structures/${structureId}/trombinoscope/reports?limit=${limit}&offset=${offset}`)
            .then((res: AxiosResponse) => {
                return res.data.all;
            });
    },

    /**
     * Clear failures history
     *
     * @param structureId   structure identifier
     */
    deleteFailuresHistory: (structureId: string): Promise<AxiosResponse> => {
        return http.delete(`/viescolaire/structures/${structureId}/trombinoscope/failures`);
    }

};

export const TrombinoscopeService = ng.service('TrombinoscopeService', (): ITrombinoscopeService => trombinoscopeService);
